shopfloor/services/delivery.py

```diff
@@ -123,16 +451,78 @@ def reset_qty_done_line(self, picking_id, line_id):
         Transitions:
         * deliver: always return here with updated data
         """
-        return self._response()
+        picking = self.env["stock.picking"].browse(picking_id).exists()
+        if picking:
+            response = self._check_picking_status(picking)
+            if response:
+                return response
+        else:
+            return self._response_for_deliver(
+                message=self.msg_store.stock_picking_not_found()
+            )
+        line = self.env["stock.move.line"].browse(move_line_id).exists()
+        if line:
+            if line.picking_id != picking:
+                return self._response_for_deliver(
+                    picking=picking,
+                    message=self.msg_store.line_not_available_in_picking(picking),
+                )
+            if line.package_id:
+                return self._response_for_deliver(
+                    picking=picking,
+                    message=self.msg_store.line_has_package_scan_package(),
+                )
+            self._reset_lines(line)
+            return self._response_for_deliver(picking)
+        return self._response_for_deliver(
+            picking=picking, message=self.msg_store.record_not_found(),
+        )
 
-    def done(self, picking_id):
+    def done(self, picking_id, confirm=False):
         """Set the stock picking to done
 
         Transitions:
         * deliver: error during action
         * confirm_done: when not all lines of the stock.picking are done
         """
-        return self._response()
+        picking = self.env["stock.picking"].browse(picking_id).exists()
+        if picking:
+            response = self._check_picking_status(picking)
+            if response:
+                return response
+        else:
+            return self._response_for_deliver(
+                message=self.msg_store.stock_picking_not_found()
+            )
+        all_qty_done = all(
+            ml.qty_done >= ml.product_uom_qty for ml in picking.move_line_ids
+        )
+        if all_qty_done:
```

Comment:
This check is done in `_action_picking_done` already, couldn't you directly call it?


```python
if self._action_picking_done(picking):
    return self._response_for_deliver(
        message=self.msg_store.transfer_complete(picking)
    )
```